### PR TITLE
feat: KG RAG query engine, enable Graph RAG on existing KGs

### PR DESCRIPTION
**Pull request by wey-gu**
# Description

- Now we could do Graph RAG towards existing KGs
- Enabled nl2cypher for older NebulaGraph, with improved schema fetching impl., reported by @pachgadehardik via [here](https://github.com/jerryjliu/llama_index/pull/2581#issuecomment-1668960096)

The shape now is RFC @logan-markewich @jerryjliu , in case good to go I'll impl. the async functions, too.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by logan-markewich — COMMENTED** [review]
I think this looks pretty good so far! Thanks for contributing this! 🙏🏻 

Just to confirm how I think this is working
- Take a query, extract out keywords and synonyms of those keywords
- Use those to extract triplets that mention these keywords/synonyms
- Optionally combine those retrieved triplets with a text2cypher query result?

**Comment by wey-gu** [comment]
> I think this looks pretty good so far! Thanks for contributing this! 🙏🏻
> 
> Just to confirm how I think this is working
> 
> * Take a query, extract out keywords and synonyms of those keywords
> * Use those to extract triplets that mention these keywords/synonyms
> * Optionally combine those retrieved triplets with a text2cypher query result?

Yes, exactly!

And for the entity/keyword search of step 1, in the future, we could introduce the embedding-based capability to enable semantic searching, too.

For step 2: extract triplets from KG, or we could say it's getting n-depth subgraphs of those entities.

And for step 3, the text2cypher could be helpful but incomplete for the retrieval, and vice visa, thus options to combine the two could be one strategy to go with on KG.

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+        self._service_context = service_context or ServiceContext.from_defaults()
+
+        # Get Graph Store Type
+        self._graph_store_type = GRAPH_STORE_CLASS_TO_GRAPH_STORE_TYPE[
+            self._graph_store.__class__
+        ]
+
+        # Get Graph schema
+        self._graph_schema = self._graph_store.get_schema()
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:122]: Sorry, in further changes, we may choose to leverage schema for certain cases, for now, it's not needed, will cleanup the unnecessary variables.

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        assert (
+            storage_context.graph_store is not None
+        ), "Must provide a graph store in the storage context."
+        self._storage_context = storage_context
+        self._graph_store = storage_context.graph_store
+
+        self._service_context = service_context or ServiceContext.from_defaults()
+
+        # Get Graph Store Type
+        self._graph_store_type = GRAPH_STORE_CLASS_TO_GRAPH_STORE_TYPE[
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:117]: Sorry, removed it.

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+            logger.info("> No knowledge sequence extracted from entities.")
+            return []
+
+        context_string = (
+            f"The following are knowledge sequence in max depth"
+            f" {self._graph_traversal_depth} "
+            f"in the form of "
+            f"`subject [predicate, object, predicate_next_hop, object_next_hop ...]`"
+            f" extracted from the query string:\n"
+            f"'\n'.join(knowledge_sequence)"
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:311]: Thanks @logan-markewich , exactly!

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+
+    def _retrieve_embedding(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
+        """retrieve in embedding mode."""
+        # TBD: will implement this later with vector store.
+        raise NotImplementedError
+
+    def _retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
+        """Build nodes for response."""
+        nodes, nodes_keyword, nodes_embedding = [], [], []
+        if self._with_nl2graphquery:
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:334]: Yes, as text2cypher are in the retrieve() phase(not explicitly implemented an as_retriever() though for now), and also exposing the retriever function could enable the granularity of a custom retriever to combine GraphRAG with any other retrievers :)

For that retrieve(_kg_query_engine._retrieve() ) calls for _service_context.llm_predictor.predict and graph_store.query(), I'll add one extra try/except on the safe side.

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        retriever_mode: Optional[str] = "keyword",
+        with_nl2graphquery: bool = False,
+        similarity_top_k: int = 5,
+        graph_traversal_depth: int = 2,
+        max_knowledge_sequence: int = 30,
+        verbose: bool = False,
+        response_synthesizer: Optional[BaseSynthesizer] = None,
+        **kwargs: Any,
+    ):
+        self._retriever = KnowledgeGraphRAGRetriever(
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:411]: Yes, the whole thing of RAG is actually a retriever, will do so with RetrieverQueryEngine!(👍🏻)

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+                storage_context=self._storage_context,
+                graph_query_synthesis_prompt=graph_query_synthesis_prompt,
+                graph_response_answer_prompt=graph_response_answer_prompt,
+                refresh_schema=refresh_schema,
+                verbose=verbose,
+                response_synthesizer=response_synthesizer,
+                **kwargs,
+            )
+
+        self._similarity_top_k = similarity_top_k
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:162]: Yes, sorry, I was planning for embedding search, but not yet implemented now, will remove

File: llama_index/query_engine/knowledge_graph_rag_query_engine.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+            )
+        )
+        return [node]
+
+    def _retrieve_embedding(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
+        """retrieve in embedding mode."""
+        # TBD: will implement this later with vector store.
+        raise NotImplementedError
+
+    def _retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
```
[the remark this answers is not in the document]
Comment (wey-gu) [reply llama_index/query_engine/knowledge_graph_rag_query_engine.py:436]: done :)

File: llama_index/graph_stores/nebulagraph.py [hunk 8]
```diff
@@ -16,11 +16,17 @@
 logger = logging.getLogger(__name__)
 
 
-rel_query = Template(
+rel_query_0 = Template(
```
Comment (logan-markewich) [inline llama_index/graph_stores/nebulagraph.py:19]: nit: can we give these a more descriptive name?

**Comment by logan-markewich** [comment]
Thanks for the updates @wey-gu !! I just made a bit of a change to move remove the RAG query engine, and just move the retriever to be used in a `RetrieverQueryEngine` (Since the RAG query engine was not doing anything that the `RetrieverQueryEngine` was not doing.

I updated the notebooks to confirm they still work as well, but you may need to update the screenshot of the streamline app :)

File: llama_index/graph_stores/nebulagraph.py (hunk 8 above)
Comment (wey-gu) [reply llama_index/graph_stores/nebulagraph.py:19]: Will do so :)

**Comment by wey-gu** [comment]
> Thanks for the updates @wey-gu !! I just made a bit of a change to move remove the RAG query engine, and just move the retriever to be used in a `RetrieverQueryEngine` (Since the RAG query engine was not doing anything that the `RetrieverQueryEngine` was not doing.
> 
> I updated the notebooks to confirm they still work as well, but you may need to update the screenshot of the streamline app :)
> 
> After that, I think it's basically good to merge!

Now it's elegant :) It should have been settled in pure retrievers!

The demo video could be as-is as it was created from the kg_index --> as query engine, but generationally enough to compare the two(text2cyper vs graph rag).

**Comment by logan-markewich** [comment]
Thanks for the amazing work on this @wey-gu 💪💪

**Comment by wey-gu** [comment]
> Thanks for the amazing work on this @wey-gu 💪💪

Thanks a ton @logan-markewich for your amazing work and support 🤩

**Comment by pachgadehardik** [comment]
Thanks a ton guys. Really appreciate the updates done in this thread.

**Comment by BleakStone** [comment]
@wey-gu 
awesome!!!
This implementation helped me a lot，but the rendered notebook [here](https://colab.research.google.com/drive/1YfJ2cPEzAJHwdDxEpUgTv4buSgrBtgNO?usp=sharing)  seems to have some version adaptation problems
Here is some of the code in rendered notebook：

```
from llama_index.query_engine import (
    KnowledgeGraphRAGRetriever,
    KnowledgeGraphRAGQueryEngine,
)

graph_rag_retriever = KnowledgeGraphRAGRetriever(
    storage_context=storage_context,
    service_context=service_context,
    llm=llm,
    verbose=True,
)

query_engine = KnowledgeGraphRAGQueryEngine(retriever=graph_rag_retriever)
```
1. I dont find the cls `KnowledgeGraphRAGQueryEngine` in llama-index==0.8.3，instead ` RetrieverQueryEngine `
```
query_engine = RetrieverQueryEngine.from_args(
    graph_rag_retriever, service_context=service_context
)
```


2. Did parameters ` llm=llm` have  effect ? And the service_context also has an llm attribute

**Comment by wey-gu** [comment]
> 

Wow, good to know it helps @BleakStone !

Sorry, we made refactor to not introduce a query engine but only a retriever(as it is only a retriever), please instead follow the final version of the notebook [here](https://gpt-index.readthedocs.io/en/stable/examples/query_engine/knowledge_graph_rag_query_engine.html) instead.

**Pull request merged**